### ISSUE TITLE
paddle.nn.Embedding和torch.nn.Embedding操作似乎不太一致

### ISSUE THREAD

**Issue by cocoshe** [PaddlePaddle/Paddle#62301]
### bug描述 Describe the Bug

+ 在paddle中，如果给Embedding设置了padding_idx，例如为0，那么可以查看他的weight，第一行全变成了0，即便是修改他的weight将第一行变成非零，由于存了padding_idx=0，[并且每次forward的时候都用这个padding_idx做了处理，直接一票否决赋0了](https://github.com/PaddlePaddle/Paddle/blob/0cb9bf687a3372cf851089fd5508f4d7fafc1295/paddle/phi/kernels/gpu/embedding_kernel.cu#L51-L52)，所以嵌入后第一行依然是全0；
```python
>>> import paddle
>>> import paddle.nn as nn
>>> x = paddle.to_tensor([0, 1, 2])
>>> vocab_size = 10
>>> dim = 3
>>> embed = nn.Embedding(vocab_size, dim)  # weight的第一行非0
>>> embed_with_padding_idx = nn.Embedding(vocab_size, dim, padding_idx=0)  # weight的第一行全为0
[2024-03-01 19:35:48,278] [ WARNING] tensor_patch_methods.py:998 - (0,)
[2024-03-01 19:35:48,278] [ WARNING] tensor_patch_methods.py:999 - 0.0
>>> embed_with_padding_idx(x)
Tensor(shape=[3, 3], dtype=float32, place=Place(cpu), stop_gradient=False,
       [[ 0.        ,  0.        ,  0.        ],
        [ 0.07028615,  0.01150006,  0.09047621],
        [-0.17829347,  0.41826904,  0.55831599]])
>>> embed_with_padding_idx.weight = embed.weight  # 这里对weight，进行修改，第一行变成了非0
>>> embed_with_padding_idx(x)  # 结果嵌入后第一行依旧是全0，这应该是由于在forward的时候用了padding_idx直接赋0了
Tensor(shape=[3, 3], dtype=float32, place=Place(cpu), stop_gradient=False,
       [[ 0.        ,  0.        ,  0.        ],
        [-0.10399878,  0.58712018,  0.46195865],
        [-0.01542068,  0.02890950, -0.43274039]])
>>> 
```

+ 在pytorch中，虽然Embedding初始化padding_idx=0，但是后续修改weight之后，嵌入得到的第一行并不是全0，应该是forward的时候没有用到，实现似乎是在[这里](https://github.com/pytorch/pytorch/blob/eee040c939ab2757f497372d28ad7696dfcc3f98/aten/src/ATen/native/Embedding.cpp#L37-L54)吗？
```python
>>> import torch
>>> import torch.nn as nn
>>> x = torch.tensor([0, 1, 2])
>>> vocab_size = 10
>>> dim = 3
>>> embed = nn.Embedding(vocab_size, dim)  # weight的第一行非0
>>> embed_with_padding_idx = nn.Embedding(vocab_size, dim, padding_idx=0)  # weight的第一行全为0
>>> embed_with_padding_idx(x)
tensor([[ 0.0000,  0.0000,  0.0000],
        [ 1.6558, -0.8115,  0.5481],
        [ 0.6362, -0.2540, -1.2049]], grad_fn=<EmbeddingBackward0>)
>>> embed_with_padding_idx.weight = embed.weight  # 修改weight，让第一行为非0
>>> embed_with_padding_idx(x)  # 虽然初始化的时候padding_idx=0，但是在torch的forward里面没有用到padding_idx这个参数
tensor([[-1.6747,  2.0626,  0.4772],
        [-1.1009, -0.1168,  0.8694],
        [ 0.0695, -0.8633, -1.8571]], grad_fn=<EmbeddingBackward0>)
>>> 
```

+ 遇到的情景：
把pytorch权重转paddle权重的时候，如果某个Embedding层的padding_idx=0，但是这一层的pretrained model给的weight第一行却不是全0，这时候即便load进参数，覆盖了这层embedding的weight，但是由于padding_idx=0在forward的时候一票否决了，所以第一行的embedding结果依然是全0；但是torch实现中覆盖后就使用覆盖后的weight，脱离了padding_idx=0的影响。

根据torch文档中的描述：https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html
![image](https://github.com/PaddlePaddle/Paddle/assets/69197635/834a79c0-9286-4332-b5bc-4d9f661071cf)
提到padding_idx可以让这一行固定为0，而且training过程并不更新，按理来说如果这个组网是train from scratch的话，那么这一行应该一直为0才对，但是我遇到的pretrained权重里面却不为0，那从文档里面`but can be updated to another value to be used as the padding vector.`，意思看样子是虽然不能自动更新，但是可以被覆盖。那我遇到的情景是不是意思可能是某一个model的组网里面Embedding初始化padding_idx=None，然后train from scratch效果不错，然后直接拿来给另一个model的Embedding做初始化，但是这个model的Embedding却是padding_idx=0。。。但是又感觉怪怪的因为我遇到的是`padding_idx=config.pad_token_id`，这样看来似乎padding_idx固定为0不被非0的pretrained weight覆盖似乎又合理一些？不知道是那篇工作的bug还是有意而为之。

**Comment by zhangbo9674** [comment]
您好，这个 API 的问题已收到，评估分析后给您答复

**Comment by zhangbo9674** [comment]
您好：确实如您所述，当前 Paddle Embedding  api 的 padding_idx 参数设置为0，会导致输出第一行全0，这一点和 torch 的 api 存在区别；如果您想 load torch 参数，可以尝试对 Paddle Embedding 不设置 padding_idx=0；对于是否要和 torch 行为对齐，需要内部对该 API 评估后给您答复，感谢您所提出的问题~

**Comment by cocoshe** [comment]
> 您好：确实如您所述，当前 Paddle Embedding api 的 padding_idx 参数设置为0，会导致输出第一行全0，这一点和 torch 的 api 存在区别；如果您想 load torch 参数，可以尝试对 Paddle Embedding 不设置 padding_idx=0；对于是否要和 torch 行为对齐，需要内部对该 API 评估后给您答复，感谢您所提出的问题~

ok明白了，辛苦啦
